discord/opus.py

```diff
@@ -276,3 +294,93 @@ def encode(self, pcm, frame_size):
             raise OpusError(ret)
 
         return array.array('b', data[:ret]).tobytes()
+
+class Decoder:
+    def __init__(self, sampling=48000, channels=2):
+        self.sampling_rate = sampling
+        self.channels = channels
+
+        self._state = self._create_state()
+
+    def __del__(self):
+        if hasattr(self, '_state'):
+            _lib.opus_decoder_destroy(self._state)
+            self._state = None
+
+    def _create_state(self):
+        ret = ctypes.c_int()
+        result = _lib.opus_decoder_create(self.sampling_rate, self.channels, ctypes.byref(ret))
+
+        if ret.value != 0:
+            log.info('error has happened in decoder state creation')
+            raise OpusError(ret.value)
+
+        return result
+
+    @classmethod
+    def _packet_get_nb_frames(cls, data):
+        """Gets the number of frames in an Opus packet"""
+        result = _lib.opus_packet_get_nb_frames(data, len(data))
+        if result < 0:
+            log.info('error has happened in packet_get_nb_frames')
+            raise OpusError(result)
+
+        return result
+
+    @classmethod
+    def _packet_get_nb_channels(cls, data):
+        """Gets the number of channels in an Opus packet"""
+        result = _lib.opus_packet_get_nb_channels(data)
+        if result < 0:
+            log.info('error has happened in packet_get_nb_channels')
+            raise OpusError(result)
+
+        return result
+
+    def _packet_get_samples_per_frame(self, data):
+        """Gets the number of samples per frame from an Opus packet"""
+        result = _lib.opus_packet_get_samples_per_frame(data, self.sampling_rate)
+        if result < 0:
+            log.info('error has happened in packet_get_samples_per_frame')
+            raise OpusError(result)
+
+        return result
+
+    def _decode(self, data, frame_size, decode_fec, decode_func, decode_ctype, arr_type):
+        if frame_size is None:
+            frames = self._packet_get_nb_frames(data)
+            samples_per_frame = self._packet_get_samples_per_frame(data)
+            # note: channels could be different from self.channels
+            # this doesn't actually get used in frame_size, but we get
+            # the value for debugging
+            channels = self._packet_get_nb_channels(data)
+            frame_size = frames * samples_per_frame
+            log.debug('detected frame_size {} ({} frames, {} samples per frame, {} channels)'.format(frame_size, frames, samples_per_frame, channels))
```

Comment:
generally you want to leave formatting for the logger (that way if it's not log level debug, the formatting is skipped)